app/code/Magento/InventoryShipping/Model/DefaultShippingAlgorithm.php

```diff
@@ -42,93 +49,174 @@ class DefaultShippingAlgorithm implements ShippingAlgorithmInterface
+                    continue;
                 }
 
-                $sourceItemQty = $sourceItem->getQuantity();
-                if ($sourceItemQty > 0) {
-                    $qtyToDeduct = min($sourceItemQty, $qtyToDeliver);
-
-                    $sourceItemSelections[$sourceItem->getSourceCode()][] = $this->sourceItemSelectionFactory->create([
+                $sourceSelections[] = $this->sourceSelectionFactory->create([
+                    'sourceCode' => $sourceItem->getSourceCode(),
+                    'sourceItemSelections' => $this->sourceItemSelectionFactory->create([
                         'sku' => $itemSku,
                         'qty' => $qtyToDeduct,
                         'qtyAvailable' => $sourceItemQty,
-                    ]);
+                    ]),
+                ]);
 
-                    $qtyToDeliver -= $qtyToDeduct;
-                }
+                $qtyToDeliver -= $qtyToDeduct;
             }
 
-            if ($qtyToDeliver > 0.0001) {
-                $this->isShippable = false;
+            if (!$this->isZero($qtyToDeliver)) {
+                $isShippable = false;
             }
         }
 
-        return $sourceItemSelections;
+        return $this->shippingAlgorithmResultFactory->create([
+            'sourceSelections' => $sourceSelections,
+            'isShippable' => $isShippable
+        ]);
+    }
+
+    /**
+     * Retrieve sources are related to current stock that are ordered by priority
+     *
+     * @param int $storeId
+     *
+     * @return SourceInterface[]
+     */
+    private function getSourcesByStoreId(int $storeId): array
+    {
+        $store = $this->storeManager->getStore($storeId);
+        $website = $this->websiteRepository->getById($store->getId());
+        $stock = $this->stockResolver->get(SalesChannelInterface::TYPE_WEBSITE, $website->getCode());
+
+       return $this->getAssignedSourcesForStock->execute($stock->getStockId());
+    }
+
+    /**
+     * Retrieve stock item from specific source by SKU
+     *
+     * @param string $stockCode
+     * @param string $sku
+     *
+     * @return SourceItemInterface
+     */
+    private function getStockItemBySku(string $stockCode, string $sku): SourceItemInterface
+    {
+        $searchCriteria = $this->searchCriteriaBuilder
+            ->addFilter(SourceItemInterface::SOURCE_CODE, $stockCode)
+            ->addFilter(SourceItemInterface::SKU, $sku)
+            ->create();
+        $sourceItemsResult = $this->sourceItemRepository->getList($searchCriteria);
+
+        if ($sourceItemsResult->getTotalCount() > 0) {
+            return reset($sourceItemsResult->getItems());
+        }
+
+        return $this->sourceItemFactory->create();
+    }
+
+    /**
+     * Compare float number with some epsilon
+     *
+     * @param float $floatNumber
+     *
+     * @return bool
+     */
+    private function isZero(float $floatNumber): bool
+    {
+        return $floatNumber < 0.0001;
```

Comment:
it's a pity that Constant has been added just in PHP 7.2 

PHP_FLOAT_EPSILON (float)
Smallest representable positive number x, so that x + 1.0 != 1.0. Available as of PHP 7.2.0.


Please, make the precision a bit less, like 0.0000001